proxylib/redis/redisparser.go

```diff
@@ -0,0 +1,447 @@
+		if len(regexStr) > 0 {
+			log.Warnf("Failed to match special-key cmd '%s' due to key-regex", cmd)
+			return false
+		}
+	case cmdNoKey:
+		break
+	}
+
+	return true
+}
+
+// RedisRuleParser parses protobuf L7 rules to enforcement objects
+// May panic
+func RedisRuleParser(rule *cilium.PortNetworkPolicyRule) []L7NetworkPolicyRule {
+	l7Rules := rule.GetL7Rules()
+	if l7Rules == nil {
+		ParseError("Can't get L7 rules.", rule)
+	}
+	var rules []L7NetworkPolicyRule
+	for _, l7Rule := range l7Rules.GetL7Rules() {
+		var rr RedisRule
+		for k, v := range l7Rule.Rule {
+			switch k {
+			case "cmd":
+				rr.cmdExact = strings.ToLower(v)
+			case "key":
+				if v != "" {
+					rr.keyRegexCompiled = regexp.MustCompile(strings.ToLower(v))
+				}
+			default:
+				ParseError(fmt.Sprintf("Unsupported key: %s", k), rule)
+			}
+		}
+
+		// ensure that 'cmd' value is a valid cmd and that key is
+		// only provided in scenarios when it makes sense.
+
+		if len(rr.cmdExact) > 0 {
+			// ensure this is a valid cmd
+			res := cmdMap[rr.cmdExact]
+			if res == cmdInvalid {
+				ParseError(fmt.Sprintf("Unable to parse L7 redis rule with invalid cmd: '%s'",
+					rr.cmdExact), rule)
+			} else if res == cmdNoKey && rr.keyRegexCompiled != nil {
+				ParseError(fmt.Sprintf("key regex not allowed for rule with cmd '%s'", rr.cmdExact), rule)
+			} else if res == cmdSpecialKey && rr.keyRegexCompiled != nil {
+				// this is temporary, as in the future we will properly parse a list of keys
+				// from these commands, and potentially allow a list of keys to match
+				ParseError(fmt.Sprintf("key regex not allowed for rule with cmd '%s'", rr.cmdExact), rule)
+			}
+
+		}
+
+		log.Debugf("Parsed RedisRule pair: %v", rr)
+		rules = append(rules, &rr)
+	}
+	return rules
+}
+
+type RedisParserFactory struct{}
+
+var redisParserFactory *RedisParserFactory
+
+func init() {
+	log.Info("init(): Registering redisParserFactory")
+	RegisterParserFactory("redis", redisParserFactory)
+	RegisterL7RuleParser("redis", RedisRuleParser)
+}
+
+type RedisParser struct {
+	connection *Connection
+	inserted   bool
+}
+
+/*
+type RedisRequestInfo struct {
+	cmd string
+	keys []string
+}
+*/
+
+const accessDeniedStr = "-Error Access Denied\r\n"
+
+func (pf *RedisParserFactory) Create(connection *Connection) Parser {
+	log.Debugf("RedisParserFactory: Create: %v", connection)
+
+	p := RedisParser{connection: connection}
+	return &p
+}
+
+func (p *RedisParser) OnData(reply, endStream bool, dataArray [][]byte) (OpType, int) {
+
+	// inefficient, but simple for now
+	data := bytes.Join(dataArray, []byte{})
+	if len(data) == 0 {
+		return MORE, 1
+	}
+
+	if reply {
+		// We currently only look at the request stream, not the reply
+		// stream.  So technically we could "PASS" with a much higher number
+		// here to improve efficiency.  At least at this point, we will keep
+		// pass the data here for visibility.
+		log.Infof("OnData: <== '%s'", data)
+		return PASS, len(data)
```

Comment:
If you are ever injecting error responses in to the reply stream then you must only `PASS` full frames here, otherwise the error responses will be injected inside of partially forwarded frames. Injections also only work when all `PASS`ed bytes have actually been forwarded, so should not `PASS` speculatively bytes before the actual frame boundaries are known.